### PR TITLE
Remove the ability to create segments in v8 format

### LINKED ISSUE #4235: Remove ability to build v8 segment format
It's possible to make segment merging more in-memory and reduce the time needed for segment merging by 50%, but it doesn't work (failing with OOM errors) when v8 segment format is build first, and then converted to v9.

I don't see a way to keep the ability to merge segments in v8 format and make the improvement without massive code duplication (because v8 and v9 use the same classes on certain level).

If we remove ability to build v8, on the contrary, some code could be removed. But not all v8-related, because segments in this format should be still readable.

### PR DESCRIPTION
**Pull request by leventov**
Fixes #4235

This PR changes the behaviour of:
 - `HadoopConverterJob`, which used to produce segments in v8 format. Now it produces segments in v9 format.
 - `ConvertSegmentTask` which delegates to `IndexIO.convertSegment()`, used to produce a segment in v8 format if given a segment in v9 format, that I believe was a bug.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 1]
```diff
@@ -115,26 +91,19 @@
...
-        .put(8, new LegacyIndexLoader(defaultIndexIOHandler, columnConfig))
-        .put(9, new V9IndexLoader(columnConfig))
-        .build();
+    ImmutableMap.Builder<Integer, IndexLoader> indexLoadersBuilder = ImmutableMap.builder();
+    LegacyIndexLoader legacyIndexLoader = new LegacyIndexLoader(new DefaultIndexIOHandler(), columnConfig);
+    for (int i = 0; i <= V8_VERSION; i++) {
+      indexLoadersBuilder.put(i, legacyIndexLoader);
+    }
+    indexLoadersBuilder.put((int) V9_VERSION, new V9IndexLoader(columnConfig));
+    indexLoaders = indexLoadersBuilder.build();
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: marker for my self LGTM to this point.

File: indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java [hunk 2]
```diff
@@ -119,7 +116,7 @@ public Plumber findPlumber(
     final Set<File> spilled = Sets.newHashSet();
 
     // IndexMerger implementation.
-    final IndexMerger theIndexMerger = config.getBuildV9Directly() ? indexMergerV9 : indexMerger;
+    final IndexMerger theIndexMerger = indexMergerV9;
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java]: The ```theIndexMerger``` variable can be simply removed.

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 3]
```diff
@@ -481,362 +408,6 @@ public MMappedIndex mapDir(File inDir) throws IOException
...
-            MetricHolder holder = MetricHolder.fromByteBuffer(v8SmooshedFiles.mapFile(filename), v8SmooshedFiles);
-            final String metric = holder.getName();
-
-            final ColumnDescriptor.Builder builder = ColumnDescriptor.builder();
-
-            switch (holder.getType()) {
-              case LONG:
-                builder.setValueType(ValueType.LONG);
-                builder.addSerde(
-                    LongGenericColumnPartSerde.legacySerializerBuilder()
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Then we do not need all those legacy Ser/Des ?

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 4]
```diff
@@ -308,752 +125,112 @@ public IndexableAdapter apply(final QueryableIndex input)
...
-      timeWriter.open();
-
-      ArrayList<MetricColumnSerializer> metWriters = Lists.newArrayListWithCapacity(mergedMetrics.size());
-      final CompressedObjectStrategy.CompressionStrategy metCompression = indexSpec.getMetricCompression();
-      final CompressionFactory.LongEncodingStrategy longEncoding = indexSpec.getLongEncoding();
-      for (String metric : mergedMetrics) {
-        ValueType type = valueTypes.get(metric);
-        switch (type) {
-          case LONG:
-            metWriters.add(new LongMetricColumnSerializer(metric, v8OutDir, ioPeon, metCompression, longEncoding));
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Same thing here do we still need `MetricColumnSerializer` ?

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 5]
```diff
@@ -308,752 +125,112 @@ public IndexableAdapter apply(final QueryableIndex input)
...
-
-      /************* Setup Dim Conversions **************/
-      progress.progress();
-      startTime = System.currentTimeMillis();
-
-      final ArrayList<FileOutputSupplier> dimOuts = Lists.newArrayListWithCapacity(mergedDimensions.size());
-      final DimensionHandler[] handlers = makeDimensionHandlers(mergedDimensions, dimCapabilities);
-      final List<DimensionMerger> mergers = new ArrayList<>();
-      for (int i = 0; i < mergedDimensions.size(); i++) {
-        DimensionMergerLegacy merger = handlers[i].makeLegacyMerger(
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: This `DimensionMergerLegacy` can go too i guess

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 6]
```diff
@@ -19,261 +19,78 @@
...
-import java.util.Map;
 import java.util.NoSuchElementException;
 import java.util.PriorityQueue;
 import java.util.Set;
+import java.util.stream.Collectors;
 
-/**
- */
-public class IndexMerger
+@ImplementedBy(IndexMergerV9.class)
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: do we expect that this class to be injected via other modules ? or it is to make code look nicer ?

**Comment by b-slim** [comment]
Overall LGTM there is some unused classes that can be removed.

File: indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java (hunk 2 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java]: Removed

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 3 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Removed

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 4 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Thanks for noticing, removed

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 6 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: It is injected somewhere in indexing already. Other implementation is expected with segment format v10.

**Comment by leventov** [comment]
@jon-wei @b-slim any more comments?

**Comment by b-slim** [comment]
👍

**Comment by jon-wei** [comment]
LGTM, :+1:

**Pull request merged**